### DISCUSSION TITLE
how to keep atomic in Base operations

### DISCUSSION THREAD

**Discussion by lihugang** [deta/cloud-docs#390] [Help] [answered]
I want to know that whether there is only one instance on Deta Micro Server. If it is true, I can keep a lock in memory.

**Comment by AyushSehrawat** [comment]
You get a incoming request on your server, the micro awakes with an execution ram of 128mb ( 128mb solely for your program), after 10s it gets deleted. It happens for each request.
Though if you want to, you can increase it

**Reply by lihugang** [reply]
Thanks a lot. And how to keep data (in Base) atomic.  
For example, an increment operation.  (such as "like" operation in social app)
There is a timeline for example:  
- [Instance A] Receiving an increment request at **0ms**   
- [Instance A] Getting value from Deta Base (I noticed that it's remote, so it will take at least 10ms) at **30ms**, value is `1`    
- [Instance B] Receiving an increment request at **40ms**  
- [Instance B] Getting value at **70ms**, value is also `1`   
- [Instance A] Putting value `2` to Deta Base at **80ms**
- [Instance B] Putting value `2` to Deta Base at **100ms**

You can see that clients sent two increment requests, and it only added 1.  
How to solve this problem

Sorry, I'm not good at English, if my expression has some problems, please forgive it

**Reply by AyushSehrawat** [reply]
Something like, if client sent submit request twice with an time gap of like few ms, you only want 1 ? Or do you mean you want to keep both ? Or do you mean like save none?

**Reply by AyushSehrawat** [reply]
![image](https://user-images.githubusercontent.com/69469790/185410567-249d0fe6-d10f-40f2-9345-f6e3384244cb.png)

**Reply by lihugang** [reply]
keep both, in one record, in one field

**Reply by AyushSehrawat** [reply]
Like, if user sends two requests within a short timeline, store both inside single record?

**Reply by lihugang** [reply]
yes

**Reply by AyushSehrawat** [reply]
Your language ?

**Reply by lihugang** [reply]
node v14

**Reply by AyushSehrawat** [reply]
I am not sure about such situations , maybe try adding checks, but i wonder if they will work within short intervals :/ .

**Reply by lihugang** [reply]
how to check? I have no ideas.

**Reply by AyushSehrawat** [reply]
The most i can think is like, instead of put , insert them both, so they don't overwrite each other, and then make some check to see if same data exists, merge them

**Reply by AyushSehrawat** [reply]
Add some cron job else, which sees database for same values and merge them , optionally you can add created at time and then check , if time difference small merge

**Reply by AyushSehrawat** [reply]
Another can be, make some background check.

When a request comes to server -> micro for that request gets created ( killed in T-10) -> database operation starts

During this 10s interval add some background task to check it.

**Reply by lihugang** [reply]
> Add some cron job else, which sees database for same values and merge them , optionally you can add created at time and then check , if time difference small merge

I think it is practicable, but not prefect. That will put too much burden on Micro and Base. Thanks a lot.

**Reply by AyushSehrawat** [reply]
cron jobs are here https://docs.deta.sh/docs/micros/cron/#deta-cron. Also don't worry bout the burden thing ;-; they get timeout anyways

**Reply by jnsougata** [reply]
I still don’t understand the issue and use case of micro here

**Reply by lihugang** [reply]
> I still don’t understand the issue and use case of micro here

Ensure that operations on Base are atomic.
That's my problem

**Reply by AyushSehrawat** [reply]
think it like a user accidently clicking submit button twice in short interval, he/she wants to save both in one record

**Reply by lihugang** [reply]
> think it like a user accidently clicking submit button twice in short interval, he/she wants to save both in one record

maybe two users clicking submit button in short interval (such as ”like" operation)

**Comment by lihugang** [comment]
Could others give some better solutions? Thanks.

**Reply by jnsougata** [reply]
> Ensure that operations on Base are atomic. That's my problem

what do you mean by atomic?

you said atomic to me which means single i.e. once clicked the data is created to base so no further data will be created
In that case  just use `insert` 

or just send `increment` `update` to existing data

**Reply by AyushSehrawat** [reply]
Its between very short intervals, wonder if it will work :thonk:

**Reply by jnsougata** [reply]
> Its between very short intervals, wonder if it will work :thonk:

ofc it will lol you don’t have to worry about interval…. there wont be anything that will create race condition to request…. the request which registers first… data will be written to base accordingly… remember writing something to memory is always sync

**Reply by AyushSehrawat** [reply]
Hm, then that's good, @lihugang just try with some conditions like, if already exists update, if it doesn't work, try our something else .

**Reply by lihugang** [reply]
I will do an experiment to check it.

**Reply by jnsougata** [reply]
> I will do an experiment to check it.

    try:
       Insert
    except:
        Increment 
something like this

**Reply by lihugang** [reply]
In fact, I want to do complex operations.
If the length of the current array is over limit (such as 50), insert a new array to Base
else push data to array.

The latter can be realized by update operation, but the first how to keep atomic

**Reply by lihugang** [reply]
[Example website](https://test0818.deta.dev/)
I sent 9 requests, but it only added 5

- - -
Now this website I deleted.

**Reply by lihugang** [reply]
@jnsougata @AyushSehrawat 
I think function `Base.get`  and `Base.put`  requests the remote server, not storing in memory

**Reply by jnsougata** [reply]
> [Example website](https://test0818.deta.dev/)
> I sent 9 requests, but it only added 5

can yo share some code?

**Reply by lihugang** [reply]
> > I will do an experiment to check it.
> 
> ```
> try:
>    Insert
> except:
>     Increment 
> ```
> 
> something like this

My requirements are too complex to be implemented with the functions provided by the sdk
I only trys the `put`

**Reply by jnsougata** [reply]
> @jnsougata @AyushSehrawat
> I think function `Base.get` and `Base.put` requests the remote server, not storing in memory

of they they don’t store in memory…. there’s nothing in memory in deta

**Reply by lihugang** [reply]
so it takes 20-200ms to access Base, is async

**Reply by lihugang** [reply]
> can you share some code

@jnsougata 
Source code
```javascript

const express = require('express');
const app = express();
const { Deta } = require('deta');
const deta = Deta('<project key>');
const base = deta.Base('test-base');
app.get('/', function (req, res) {
    res.writeHead(200, {
        'Content-type': 'text/html'
    });
    res.end(`
    <button onclick=javascript:send_request(9);>Click</button> the button and open f12 devtools to watch requests
    <script>
    function send_request(remain_count) {
        if (remain_count <= 0) return;
        fetch('/', {
            method: 'POST'
        });
        if (Math.random() > 0.5)
            setTimeout(send_request.bind(this, remain_count - 1), parseInt(100 * Math.random()));
        else send_request(remain_count - 1);
    };
    //it will send 9 requests, 50% chance send immediately
    //50% chance send in 0-100ms
    </script>
    `);
});
app.post('/', async function (req, res) {
    var count = await base.get('count');
    if (!count) count = 0;
    else count = count.count;
    count++;
    await base.put({
        count: count
    }, 'count');
    res.writeHead(200);
    res.write(count.toString());
    res.end();
});
module.exports = app;
```

**Reply by jnsougata** [reply]
> > can you share some code
> 
> @jnsougata Source code
> 
> ```js
> const express = require('express');
> const app = express();
> const { Deta } = require('deta');
> const deta = Deta('<project key>');
> const base = deta.Base('test-base');
> app.get('/', function (req, res) {
>     res.writeHead(200, {
>         'Content-type': 'text/html'
>     });
>     res.end(`
>     <button onclick=javascript:send_request(9);>Click</button> the button and open f12 devtools to watch requests
>     <script>
>     function send_request(remain_count) {
>         if (remain_count <= 0) return;
>         fetch('/', {
>             method: 'POST'
>         });
>         if (Math.random() > 0.5)
>             setTimeout(send_request.bind(this, remain_count - 1), parseInt(100 * Math.random()));
>         else send_request(remain_count - 1);
>     };
>     //it will send 9 requests, 50% chance send immediately
>     //50% chance send in 0-100ms
>     </script>
>     `);
> });
> app.post('/', async function (req, res) {
>     var count = await base.get('count');
>     if (!count) count = 0;
>     else count = count.count;
>     count++;
>     await base.put({
>         count: count
>     }, 'count');
>     res.writeHead(200);
>     res.write(count.toString());
>     res.end();
> });
> module.exports = app;
> ```

don't use `put` it will overwrite the existing data.. use `insert` and use a proper key to guarantee the `update` or `insert` data

**Reply by lihugang** [reply]
![image](https://user-images.githubusercontent.com/94808542/185427667-267a9509-43af-440d-8501-dca19e21aa39.png)
`insert` will create a new record, I would like to update fields in one record

**Reply by jnsougata** [reply]
> ![image](https://user-images.githubusercontent.com/94808542/185427667-267a9509-43af-440d-8501-dca19e21aa39.png) `insert` will create a new record, I would like to update fields in one record

that's why i told you to use a proper `key` with `data`
![image](https://user-images.githubusercontent.com/53375272/185428283-0c355cf1-febd-4e07-9ec9-acc4ae63bdf8.png)

**Reply by lihugang** [reply]
such as `count.<timestamp>`?

**Reply by lihugang** [reply]
and merge them sometime?

**Reply by jnsougata** [reply]
> such as `count.<timestamp>`?

any key which you will use as reference while sending an update or inserting a value `abc` `xyz` `pqr` ....

**Reply by lihugang** [reply]
it's the same with AyushSehrawat said

**Reply by jnsougata** [reply]
> it's the same with AyushSehrawat said
🤯 manh! I questioned the role of corn here... not who said what! cause you don't need corn in this context

**Reply by lihugang** [reply]
so...... no merge, when quering, fetching data by `contains` parameter?

**Reply by jnsougata** [reply]
> so...... no merge, when quering, fetching data by `contains` parameter?

just use `get`

**Reply by lihugang** [reply]
> just use `get`

As you said, there are some random values in key such as `abc` `xyz` `pqr`, how could I know the correct key?

**Reply by AyushSehrawat** [reply]
bruh, i guess just put the value into the database with a key

![image](https://user-images.githubusercontent.com/69469790/185431338-dc556654-b2b6-45c6-ac9f-d8875c47085d.png)


And make a check when the post request is sent , whether this key exists or not, if it exists, update that else create

**Reply by AyushSehrawat** [reply]
> > just use `get`
> 
> As you said, there are some random values in key such as `abc` `xyz` `pqr`, how could I know the correct key?

You can set the key

**Reply by AyushSehrawat** [reply]
![image](https://user-images.githubusercontent.com/69469790/185431975-8986ea05-6000-4f75-bcf9-28a0fc4465ab.png)

If no key provided it randomly creates one

**Reply by jnsougata** [reply]
> > just use `get`
> 
> As you said, there are some random values in key such as `abc` `xyz` `pqr`, how could I know the correct key?

`db.insert(key=your_known_key, blah blah)`
`db.get(your_known_key)`

**Reply by jnsougata** [reply]
> bruh, i guess just put the value into the database with a key
> 
> ![image](https://user-images.githubusercontent.com/69469790/185431338-dc556654-b2b6-45c6-ac9f-d8875c47085d.png)
> 
> And make a check when the post request is sent , whether this key exists or not, if it exists, update that else create

`insert` exists

**Reply by AyushSehrawat** [reply]
@jnsougata He can use `put` na, you said operations will handle it, so if he puts like at 10ms and another request at 12ms for put, before putting just make a check by get

**Reply by AyushSehrawat** [reply]
`insert` destroys his/her whole purpose of maintaining only 1 record

**Reply by jnsougata** [reply]
> @jnsougata He can use `put` na, you said operations will handle it, so if he puts like at 10s and another request at 12 second for put, before putting just make a check by get

don't give him wrong ideas... put will just overwrite existing data moreover why you will use `put` and check where you can just ust `insert`

**Reply by AyushSehrawat** [reply]
~~at this point chat feels messed up lol~~

@lihugang Just try setting key and do your operations

**Reply by jnsougata** [reply]
> `insert` destroys his/her whole purpose of maintaining only 1 record

that's why providing shit_known_key_is_recommended

**Reply by AyushSehrawat** [reply]
> > @jnsougata He can use `put` na, you said operations will handle it, so if he puts like at 10s and another request at 12 second for put, before putting just make a check by get
> 
> don't give him wrong ideas... put will just overwrite existing data moreover why you will use `put` and check where you can just ust `insert`

Bruh i mean, 

click submit -> request goes to post endpoint -> deta base operation takes place

In deta base operations add a check for first get, if return nothing , then put else update

**Reply by jnsougata** [reply]
> 

insert will raise if the key already exist where put won't @AyushSehrawat 
now:
```
  req1:
      sent a put
  req 2:
      sent a put
   
  it won't even care which req registers 1st
 ----
 req1:
      sent an insert
  req 2:
      sent an insert

   whatever req registered first will insert the key hence 
  any other req registered after the 1st request can not insert the same key 
  hence it will go for update-increment
  ```

**Reply by AyushSehrawat** [reply]
> > 
> 
> insert will raise if the key already exist where put won't @AyushSehrawat 
> now:
> ```
>   req1:
>       sent a put
>   req 2:
>       sent a put
>    
>   it won't even care which req registers 1st
>  ----
>  req1:
>       sent an insert
>   req 2:
>       sent an insert
> 
>    whatever req registered first will insert the key hence 
>   any other req registered after the 1st request can not insert the same key 
>   hence it will go for update-increment
>   ```
>  

Bruh did you read what i sent. First get, to see if it exists, if not put, else update

**Reply by jnsougata** [reply]
> > > 
> > 
> > 
> > insert will raise if the key already exist where put won't @AyushSehrawat
> > now:
> > ```
> >   req1:
> >       sent a put
> >   req 2:
> >       sent a put
> >    
> >   it won't even care which req registers 1st
> >  ----
> >  req1:
> >       sent an insert
> >   req 2:
> >       sent an insert
> > 
> >    whatever req registered first will insert the key hence 
> >   any other req registered after the 1st request can not insert the same key 
> >   hence it will go for update-increment
> > ```
> 
> Bruh did you read what i sent. First get, to see if it exists, if not put, else update

bruh!!!!!!! why one would prefer sending `n` request rather than sending `n-1` requests

**Reply by AyushSehrawat** [reply]
> > > > 
> > > 
> > > 
> > > insert will raise if the key already exist where put won't @AyushSehrawat
> > > now:
> > > ```
> > >   req1:
> > >       sent a put
> > >   req 2:
> > >       sent a put
> > >    
> > >   it won't even care which req registers 1st
> > >  ----
> > >  req1:
> > >       sent an insert
> > >   req 2:
> > >       sent an insert
> > > 
> > >    whatever req registered first will insert the key hence 
> > >   any other req registered after the 1st request can not insert the same key 
> > >   hence it will go for update-increment
> > > ```
> > 
> > Bruh did you read what i sent. First get, to see if it exists, if not put, else update
> 
> bruh!!!!!!! why one would prefer sending `n` request rather than sending `n-1` requests

He/She wants 1 record with all requests that's why....

Wait, idk maybe i took something else... I thought he wants to store all contents :/..... Sorry bout that 🥲

**Reply by lihugang** [reply]
> `db.insert(key=your_known_key, blah blah)` `db.get(your_known_key)`
> that's why i told you to use a proper key with data

emm....... I think I didn't understand it well.  
That's what I understanded  
`a proper key` is add some meta data like timestamp to the original key (Like `data.<timestamp>`) to prevent repeating
req: 1  
 -   insert key=data.&lt;timestamp1&gt;   
  
req: 2  
 -   insert key=data.&lt;timestamp2&gt;    
 
And when I am quering, I don't know the timestamp I inserted, how to get data

**Comment by jnsougata** [comment]
@lihugang are you trying to make click tracker or something?

**Reply by lihugang** [reply]
In fact, I would like to store users' comments, and pages in Base to make quering operation more convenient.  
For example:  
One page has 5 records  

| key | comments | nextCursor |
------|-------------|----------------------
| cursor.x | ['xxx', 'xxx', 'xxx', 'xxx', 'xxx'] | cursor.y |
| cursor.y | ['xxx', 'xxx'] | null |

Sending hundreds of records in one response would be a huge burden on the clients.  
And using `fetch` + `limit` would take more time to get, and would also be a huge burden on `Base`

**Comment by lihugang** [comment, marked as answer]
I have a idea.  
request 1:
- Insert `key=lock` to Base  (Get lock)  
- Insert successfully  
- Update data
- Delete `key=lock` on Base (Release lock)
- End

request 2:
- Insert `key=lock` to Base (Get lock)  
- Failed to insert
- Trying to insert lock in 50ms until insert successfully
- Update data
- Release lock
- End

I think I can set `expiresIn` to the lock and it will be automatically released in 10s to solve deadlock problem.

**Reply by lihugang** [reply]
Is it the same as @jnsougata you said?

**Reply by jnsougata** [reply]
man! I'll just leave it here! You are overthinking it...

**Reply by lihugang** [reply]
ok, thank you
